accord-core/src/main/java/accord/coordinate/Recover.java

```diff
@@ -244,17 +212,57 @@ public void onSuccess(Id from, RecoverReply reply)
             recover();
     }
 
+    BiConsumer<? super Outcome, Throwable> finishAndTakeCallback()
+    {
+        return finishAndWrapCallback();
+    }
+
+    private BiConsumer<? super Outcome, Throwable> finishAndWrapCallback()
+    {
+        BiConsumer<? super Outcome, Throwable> callback = super.finishAndTakeCallback();
+        return (success, failure) -> {
+            callback.accept(success, failure);
+            node.agent().coordinatorEvents().onRecoveryStopped(node, txnId, ballot, null, failure);
+        };
+    }
+
+    private BiConsumer<? super Outcome, Throwable> finishAndUnwrapCallback()
+    {
+        return super.finishAndTakeCallback();
+    }
+
+    private BiConsumer<Result, Throwable> finishAndTakeResultCallback()
+    {
+        BiConsumer<? super Outcome, Throwable> callback = finishAndUnwrapCallback();
+        return (success, failure) -> {
+            if (failure == null)
+            {
+                callback.accept(ProgressToken.APPLIED, null);
+                node.agent().coordinatorEvents().onRecoveryStopped(node, txnId, ballot, success, failure);
+            }
+            else if (failure instanceof Redundant)
+            {
+                Timestamp committedExecuteAt = ((Redundant) failure).committedExecuteAt;
+                if (tracing != null)
+                    tracing.trace(null, "Recover found Redundant; retrying with known committedExecuteAt " + committedExecuteAt);
+                retry(committedExecuteAt, callback);
```

Comment:
Could you add an invariant that current recover is done before calling retry? Maybe superfluous.